File: backend-spring/src/main/java/com/nearkart/entity/FarmerSale.java
package com.nearkart.entity;

import jakarta.persistence.*;
import lombok.*;
import java.math.BigDecimal;
import java.time.OffsetDateTime;
import java.util.UUID;

@Entity
@Table(name = "farmer_sales")
@Getter @Setter
@NoArgsConstructor @AllArgsConstructor
@Builder
public class FarmerSale {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @Column(name = "user_id", nullable = false)
    private UUID userId;

    @Column(name = "company_id")
    private UUID companyId;

    @Column(name = "farmer_entry_id", nullable = false)
    private UUID farmerEntryId;

    @Column(name = "market_fees", precision = 14, scale = 2)
    @Builder.Default
    private BigDecimal marketFees = BigDecimal.ZERO;

    @Column(precision = 14, scale = 2)
    @Builder.Default
    private BigDecimal supervision = BigDecimal.ZERO;

    @Column(name = "adat_commission", precision = 14, scale = 2)
    @Builder.Default
    private BigDecimal adatCommission = BigDecimal.ZERO;

    @Column(precision = 14, scale = 2)
    @Builder.Default
    private BigDecimal bardan = BigDecimal.ZERO;

    @Column(precision = 14, scale = 2)
    @Builder.Default
    private BigDecimal labour = BigDecimal.ZERO;

    @Column(precision = 14, scale = 2)
    @Builder.Default
    private BigDecimal sutli = BigDecimal.ZERO;

    @Column(name = "gadi_bhada", precision = 14, scale = 2)
    @Builder.Default
    private BigDecimal gadiBhada = BigDecimal.ZERO;

    @Column(name = "weight_short", precision = 14, scale = 2)
    @Builder.Default
    private BigDecimal weightShort = BigDecimal.ZERO;

    @Column(name = "total_deductions", precision = 14, scale = 2)
    @Builder.Default
    private BigDecimal totalDeductions = BigDecimal.ZERO;

    @Column(name = "net_payable", precision = 14, scale = 2)
    @Builder.Default
    private BigDecimal netPayable = BigDecimal.ZERO;

    @Column(name = "created_at", nullable = false, updatable = false)
    private OffsetDateTime createdAt;

    @Column(name = "updated_at", nullable = false)
    private OffsetDateTime updatedAt;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "farmer_entry_id", insertable = false, updatable = false)
    private FarmerEntry farmerEntry;

    @PrePersist
    protected void onCreate() { OffsetDateTime now = OffsetDateTime.now(); createdAt = now; updatedAt = now; }

    @PreUpdate
    protected void onUpdate() { updatedAt = OffsetDateTime.now(); }
}

File: backend/app/schemas/farmer_sale.py
from datetime import datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, ConfigDict


class FarmerSaleCreate(BaseModel):
    company_id: UUID | None = None
    farmer_entry_id: UUID
    market_fees: Decimal = Decimal("0")
    supervision: Decimal = Decimal("0")
    adat_commission: Decimal = Decimal("0")
    bardan: Decimal = Decimal("0")
    labour: Decimal = Decimal("0")
    sutli: Decimal = Decimal("0")
    gadi_bhada: Decimal = Decimal("0")
    weight_short: Decimal = Decimal("0")
    total_deductions: Decimal = Decimal("0")
    net_payable: Decimal = Decimal("0")


class FarmerSaleUpdate(BaseModel):
    company_id: UUID | None = None
    farmer_entry_id: UUID | None = None
    market_fees: Decimal | None = None
    supervision: Decimal | None = None
    adat_commission: Decimal | None = None
    bardan: Decimal | None = None
    labour: Decimal | None = None
    sutli: Decimal | None = None
    gadi_bhada: Decimal | None = None
    weight_short: Decimal | None = None
    total_deductions: Decimal | None = None
    net_payable: Decimal | None = None


class FarmerSaleResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    user_id: UUID
    company_id: UUID | None
    farmer_entry_id: UUID
    market_fees: Decimal
    supervision: Decimal
    adat_commission: Decimal
    bardan: Decimal
    labour: Decimal
    sutli: Decimal
    gadi_bhada: Decimal
    weight_short: Decimal
    total_deductions: Decimal
    net_payable: Decimal
    created_at: datetime
    updated_at: datetime
